### PR TITLE
Benchmem

### PR DESCRIPTION
**Pull request by tgulacsi**
See https://github.com/shamaazi/antilog/commit/80df4e49ecc3b2921a57944637b3bb4f25c549c1 for the benchmarks.

Although Go is fast, logging should not have a big performance/memory penalty - or at least it should consume as less resources as possible.

### CODE REVIEW

File: bench_lookup_test.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// +build lookup
+
+package antilog_test
```
Comment (dglsparsons) [inline bench_lookup_test.go]: This isn't testing anything antilog related. Could it be removed?

File: fields.go [hunk 2]
```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
...
-			f = append(EncodedFields{field}, f...)
+		if found {
+			continue
 		}
+		// res contains the EncodedFields, starting at offset
+		if res == nil {
+			length := len(f)
+			offset = len(fields)
+			res = make(EncodedFields, offset+length)
+			copy(res[offset:], f)
```
Comment (dglsparsons) [inline fields.go]: I think you could reassign res[offset:] to `f` here, avoiding the `res == nil` check below, meaning you can always return `f`

File: fields.go [hunk 3]
```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
 
 // PrependUnique encoded field if the key is not already set
 func (f EncodedFields) PrependUnique(fields EncodedFields) EncodedFields {
+	var res EncodedFields
+	var offset int
 	for ix := len(fields) - 1; ix >= 0; ix-- {
 		field := fields[ix]
-
+		key := field.Key()
```
Comment (dglsparsons) [inline fields.go:30]: This looks like an unnecessary assignment

File: fields.go [hunk 4]
```diff
@@ -21,21 +21,38 @@ type EncodedFields []EncodedField
 
 // PrependUnique encoded field if the key is not already set
 func (f EncodedFields) PrependUnique(fields EncodedFields) EncodedFields {
+	var res EncodedFields
+	var offset int
```
Comment (dglsparsons) [inline fields.go:27]: You could initialise `offset` here to `len(fields)`

File: log.go [hunk 5]
```diff
@@ -50,21 +50,29 @@ func (a AntiLog) Write(msg string, fields ...Field) {
...
-
-	if a.Writer == nil {
-		a.Writer = os.Stderr
+	w := a.Writer
+	if w == nil {
+		w = os.Stderr
 	}
-
-	fmt.Fprintln(a.Writer, encodedFieldsToJSONObject(encodedFields))
+	w.Write(sb.Bytes())
```
Comment (dglsparsons) [inline log.go]: You should probably dogsled this, as it has return values. 
```
_, _ = w.Write(...)
```

Comment (dglsparsons) [reply log.go]: (so it's explicit the error is being ignored)

File: log.go [hunk 6]
```diff
@@ -50,21 +50,29 @@ func (a AntiLog) Write(msg string, fields ...Field) {
...
 		PrependUnique(a.Fields)
 
-	for ix := 0; ix < len(encodedFields); ix++ {
-		field := encodedFields[ix]
-		if field.Key() != "message" || field.Key() != "timestamp" {
+	var sb bytes.Buffer
+	sb.WriteString(`{ "timestamp": "`)
+	sb.WriteString(now.Format(time.RFC3339))
+	sb.WriteString(`", "message": `)
+	json.NewEncoder(&sb).Encode(msg)
```
Comment (dglsparsons) [inline log.go]: As below - should probably explicitly ignore the error here 
```
_ = json.New...
```

**Review by dglsparsons — CHANGES_REQUESTED** [review]

File: fields.go (hunk 2 above)
Comment (dglsparsons) [reply fields.go]: It should also remove the need to assign `f` or `res` to `flds`

File: benchmark_test.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
+package antilog
```
Comment (dglsparsons) [inline benchmark_test.go]: can we put this into a `antilog_test` package instead

File: fields.go (hunk 3 above)
Comment (tgulacsi) [reply fields.go:30]: It can be moved down to line 32, but it is used in the search loop.

Comment (dglsparsons) [reply fields.go:30]: yeah - it's assigned to `key` and only used in the one place i mean, and i don't think it aids the readability over just using `if v.Key() == field.Key()` directly.

Comment (tgulacsi) [reply fields.go:30]: It is given it's own variable to not have to call Key method on every iteration of the search loop.

File: fields.go (hunk 2 above)
Comment (tgulacsi) [reply fields.go]: Godd catch, thanks for the idea!

**Review by dglsparsons — APPROVED** [review]

**Comment by dglsparsons** [comment]
Thanks for your contribution here - looks like some crazy reduction in allocations. It's much appreciated.

**Pull request merged**